contracts-bedrock: refactor Initializable.t.sol

0xfuturistic opened the pull request:
There isn't a good way to handle the implementations for the Predeploys. `Initializable.t.sol` currently uses `mustGetAddress` function. However, it is very hacky to handle all the Predeploys this way and has a big if else statement. This PR addresses this issue.

0xfuturistic committed:
contracts-bedrock: update Deployer's loadInitializedSlot

if contract is a proxy, it now gets the implementation address using either EIP1967Helper or AddressManager (for ResolvedDelegateProxy)

0xfuturistic committed:
contracts-bedrock: drop isProxy argument of loadInitializedSlot

0xfuturistic committed:
contracts-bedrock: improve logic in loadInitializedSlot

maurelian commented on packages/contracts-bedrock/scripts/Deployer.sol:
```diff
@@ -561,15 +563,44 @@ abstract contract Deployer is Script {
         initialized_ = uint8((uint256(slotVal) >> (slot.offset * 8)) & 0xFF);
     }
 
+    function _substring(
```
I'm surprised we don't already have this functionality somewhere. 
@mds1 I looked in forge-std, but couldn't find it there, did I miss something? 

I think it'd be worth either: 
1. starting a new file: `src/libraries/Strings.sol`
2. contributing this to `forge-std`

mds1 replied:
I'd prefer not to add this to forge-std as the goal is to trim it down significantly (see https://github.com/foundry-rs/foundry/issues/3782). I'd suggest just using solady's [`LibString.slice`](https://github.com/Vectorized/solady/blob/ac6d2cc13807e7e30b3c689af900d01150b03534/src/utils/LibString.sol#L699) method instead of rolling our own, since it's thoroughly tested and also gives us various other string utils

tynes replied:
I agree on using solady here, we should add it as a dep and then we can remove our vendor of solady sort

maurelian replied:
I think it is a dep?

mds1 replied:
(just unresolved this comment so it doesn't get lost, since I didn't see this change introduced yet)

0xfuturistic replied:
had locally made the commit but didn't push it. plz check now!

mds1 replied:
Looks good now, thanks! Only comment would be to pin to the latest tag, (`v0.0.158`, commit 502cc1e) instead of the most recent commit (currently using ac6d2cc). 

You can check the current version we're using with `git submodule status`, and can install a tag with `forge install https://github.com/Vectorized/solady@v0.0.158` (though I thought `forge install https://github.com/Vectorized/solady` would do this by default)

0xfuturistic replied:
good idea. thank you!

maurelian commented on packages/contracts-bedrock/src/L1/ResourceMetering.sol:
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: MIT
-pragma solidity 0.8.15;
+pragma solidity ^0.8.15;
```
Are these pragma changes intentional/necessary?

tynes replied:
I don't believe we want to change the pragma changes here altho resource metering prob could be `^0.8.0` but we shouldn't change that as part of this

0xfuturistic replied:
unfortunately these version changes are intentional, since they are dependencies of contract `EIP1967Helper`, which we are newly importing in `Deployer.sol` to use the library in `loadInitializedSlot`: https://github.com/ethereum-optimism/optimism/blob/85b5949325acc4dd7c7e85531de8336d70ae0c81/packages/contracts-bedrock/scripts/Deployer.sol#L10

If we leave these dependencies, such as `ResourceMetering.sol`, unchanged, we run into the following error at compilation:

```
[⠃] Compiling...2024-01-08T22:10:45.401194Z ERROR foundry_compilers::resolver: failed to resolve versions
[⠊] Compiling...
Error: 
Found incompatible Solidity versions:
scripts/DeployL2.s.sol (^0.8.0) imports:
    scripts/Deployer.sol (^0.8.0)
    scripts/DeployConfig.s.sol (^0.8.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    src/EAS/EAS.sol (0.8.19)
    src/EAS/SchemaRegistry.sol (0.8.19)
    src/EAS/ISchemaRegistry.sol (^0.8.0)
    src/libraries/Predeploys.sol (^0.8.0)
    lib/forge-std/src/Script.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdJson.sol (>=0.6.0 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    scripts/Executables.sol (^0.8.0)
    scripts/Chains.sol (^0.8.0)
    src/libraries/Predeploys.sol (^0.8.0)
    test/mocks/EIP1967Helper.sol (^0.8.0)
    src/legacy/AddressManager.sol (^0.8.15)
    lib/solady/src/utils/LibString.sol (^0.8.4)
    lib/forge-std/src/Script.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/StdJson.sol (>=0.6.0 <0.9.0)
    scripts/Executables.sol (^0.8.0)
    scripts/Chains.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/Address.sol (^0.8.1)
    src/universal/ISemver.sol (^0.8.0)
    src/libraries/Predeploys.sol (^0.8.0)
    src/EAS/eip1271/EIP1271Verifier.sol (0.8.19)
    src/EAS/resolver/ISchemaResolver.sol (^0.8.0)
    src/EAS/Common.sol (^0.8.0)
    src/EAS/IEAS.sol (^0.8.0)
    src/EAS/ISchemaRegistry.sol (^0.8.0)
    src/universal/ISemver.sol (^0.8.0)
    src/EAS/resolver/ISchemaResolver.sol (^0.8.0)
    src/EAS/Common.sol (^0.8.0)
    src/EAS/ISchemaRegistry.sol (^0.8.0)
    src/EAS/resolver/ISchemaResolver.sol (^0.8.0)
    lib/forge-std/src/console.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/safeconsole.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdChains.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdCheats.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdJson.sol (>=0.6.0 <0.9.0)
    lib/forge-std/src/StdMath.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStyle.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/StdUtils.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Base.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    src/libraries/Constants.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/access/Ownable.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/cryptography/draft-EIP712.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/cryptography/SignatureChecker.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/Address.sol (^0.8.1)
    src/EAS/IEAS.sol (^0.8.0)
    src/EAS/Common.sol (^0.8.0)
    src/EAS/Common.sol (^0.8.0)
    src/EAS/ISchemaRegistry.sol (^0.8.0)
    src/EAS/Common.sol (^0.8.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/console2.sol (>=0.4.22 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/interfaces/IMulticall3.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/mocks/MockERC20.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/mocks/MockERC721.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/StdStorage.sol (>=0.6.2 <0.9.0)
    lib/forge-std/src/Vm.sol (>=0.6.2 <0.9.0)
    src/L1/ResourceMetering.sol (0.8.15)
    lib/openzeppelin-contracts/contracts/utils/Context.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/Address.sol (^0.8.1)
    lib/openzeppelin-contracts/contracts/interfaces/IERC1271.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/proxy/utils/Initializable.sol (^0.8.2)
    lib/openzeppelin-contracts/contracts/utils/math/Math.sol (^0.8.0)
    src/libraries/Burn.sol (^0.8.15)
    src/libraries/Arithmetic.sol (^0.8.15)
    lib/openzeppelin-contracts/contracts/utils/Strings.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/Address.sol (^0.8.1)
    lib/openzeppelin-contracts/contracts/utils/math/SignedMath.sol (^0.8.0)
    lib/solmate/src/utils/FixedPointMathLib.sol (>=0.8.0)
 ELIFECYCLE  Test failed. See above for more details.
```

have any thoughts?

0xfuturistic replied:
ended up dropping the `Constants` import from `EIP1967Helper` that was causing the conflict and defining the relevant constants directly in the contract. this worked. feel free to take a look! https://github.com/ethereum-optimism/optimism/blob/193d8b67afc3fa27a5adebdc30e7fb06604cd195/packages/contracts-bedrock/test/mocks/EIP1967Helper.sol#L1-L27

tynes replied:
This is a fine solution for now

mds1 commented on packages/contracts-bedrock/scripts/Deployer.sol:
```diff
@@ -561,15 +563,44 @@ abstract contract Deployer is Script {
+    function loadInitializedSlot(string memory _contractName) public returns (uint8 initialized_) {
+        uint256 contractNameLength = bytes(_contractName).length;
+        address addr;
+        // Check if the contract name ends with `Proxy` and if so, get the implementation address
+        string memory suffix = _substring(_contractName, contractNameLength - 5, contractNameLength);
+        if (bytes(suffix).length == bytes("Proxy").length && keccak256(bytes(suffix)) == keccak256(bytes("Proxy"))) {
```
We now have an implicit requirement that proxy contracts must be named `*Proxy`. We should make this requirement explicit by documenting it in the style guide, and ideally trying to automate this check (e.g. semgrep rule that if `Proxy` is found in a contract name it must be the suffix) but that is tricky to do generically

tynes replied:
We already have this implicit requirement elsewhere in the codebase fwiw

mds1 replied:
Ah thanks, I didn't know that. Still agree with my recommendations, but happy to have them done in a separate PR

tynes replied:
I agree with your recommendation as well, making it explicit is better

0xfuturistic committed:
contracts-bedrock: add tests in Initializable.t.sol for L1 implementations

0xfuturistic committed:
contracts-bedrock: update logic in Initializable.t.sol to count implementations in assertion

0xfuturistic committed:
forge install: solady

v0.0.158

0xfuturistic committed:
contracts-bedrock: add solady to foundry.toml remappings

0xfuturistic committed:
contracts-bedrock: drop vendor of solady sort

0xfuturistic committed:
contracts-bedrock: use solady LibString in loadInitializedSlot

mds1 commented on packages/contracts-bedrock/scripts/Deployer.sol:
```diff
@@ -561,12 +564,23 @@ abstract contract Deployer is Script {
+            // AddressManager, which would work if the proxy is ResolvedDelegateProxy like L1CrossDomainMessengerProxy.
+            if (contractAddress == address(0)) {
+                contractAddress =
+                    AddressManager(mustGetAddress("AddressManager")).getAddress(LibString.concat("OVM_", _contractName));
```
nit, can use native solidity here
```suggestion
                    AddressManager(mustGetAddress("AddressManager")).getAddress(string.concat("OVM_", _contractName));
```

0xfuturistic replied:
this was originally the case but switched to solady's due to gas efficiency, given that the library is a dep now. should we switch back to native solidity here?

mds1 replied:
IMO we should, since this is a script we don't care about gas efficiency, so better to keep code simple and minimize deps when there's native solidity support. But this is definitely a nit

0xfuturistic replied:
I agree -- will update. thanks!

0xfuturistic committed:
contracts-bedrock: define constants in EIP1967Helper and drop Constants import

this allows us to not have to update the version of Arithmetic, ResourceMetering, and Burn, which would otherwise clash with Deployer.sol

0xfuturistic committed:
contracts-bedrock: use interface for AddressManager instead of updating version directly

0xfuturistic committed:
contracts-bedrock: use native solidity concat in loadInitializedSlot

0xfuturistic commented on packages/contracts-bedrock/test/mocks/EIP1967Helper.sol:
this change was made because the solidity versions of deps of `Constants` clash with versions of `Deployer.sol` and contracts that import the latter when `EIP1967Helper` is imported by `Deployer.sol`. thus, a temporary solution is to drop the import of `Constants` in `EIP1967Helper` and define the relevant constants locally

0xfuturistic commented on packages/contracts-bedrock/test/mocks/EIP1967Helper.sol:
```diff
@@ -2,18 +2,26 @@
 /// @title EIP1967Helper
 /// @dev Testing library to help with reading EIP 1967 variables from state
 library EIP1967Helper {
+    /// @notice The storage slot that holds the address of a proxy implementation.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1)`
+    bytes32 internal constant PROXY_IMPLEMENTATION_ADDRESS =
+        0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+
+    /// @notice The storage slot that holds the address of the owner.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.admin')) - 1)`
+    bytes32 internal constant PROXY_OWNER_ADDRESS = 0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103;
+
```
these changes were made because the versions of deps of `Constants` clash with versions of `Deployer.sol` and contracts that import the latter, which arise when `EIP1967Helper` is imported by `Deployer.sol`. thus, a temporary solution is to drop the import of `Constants` in `EIP1967Helper` and define the relevant constants locally

0xfuturistic replied:
could there be a better way to handle this? 🤔 @tynes

tynes reviewed (approved):
Great PR :D
